### PR TITLE
Task 05 sysfs implementation

### PR DESCRIPTION
**Pull request by DrCaH4ec**
Have done task №5. Implemented own class in sysfs.

### CODE REVIEW

File: 05_debug/sysfs/task_05_sysfs.c [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/* sysfs show() method. Calls the show() method corresponding to the individual sysfs file */
+static ssize_t total_calls_show(struct class *class, struct class_attribute *attr, char *buf)
+{
+	sprintf(buf, "%d\n", total_calls);
+
+	return strlen(buf);
```
Comment (ekovalyov) [inline 05_debug/sysfs/task_05_sysfs.c]: sprintf returns the length printed text. So calling strlen is redundant.

File: 05_debug/sysfs/task_05_sysfs.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/* sysfs show() method. Calls the show() method corresponding to the individual sysfs file */
+static ssize_t total_characters_processed_show(struct class *class, struct class_attribute *attr, char *buf)
+{
+	sprintf(buf, "%d\n", total_characters_processed);
+
+	return strlen(buf);
```
Comment (ekovalyov) [inline 05_debug/sysfs/task_05_sysfs.c]: The same issue as in lines 131-133.

File: 05_debug/sysfs/task_05_sysfs.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/* sysfs show() method. Calls the show() method corresponding to the individual sysfs file */
+static ssize_t characters_converted_show(struct class *class, struct class_attribute *attr, char *buf)
+{
+	sprintf(buf, "%d\n", characters_converted);
+
+	return strlen(buf);
```
Comment (ekovalyov) [inline 05_debug/sysfs/task_05_sysfs.c]: The same issue.

File: 05_debug/sysfs/test.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+echo "cat /sys/class/task_05/text"
+cat /sys/class/task_05/text
+
+echo "cat /sys/class/task_05/characters_converted"
+cat /sys/class/task_05/characters_converted
+
+echo "cat /sys/class/task_05/total_characters_processed"
+cat /sys/class/task_05/total_characters_processed
+
```
Comment (ekovalyov) [inline 05_debug/sysfs/test.sh]: Duplicated values in echo and commands. For debugging You can use -x option of sh:
#!/bin/sh -x
Another alternative is to create shell function which prints its argument and evaluate it.

File: 05_debug/sysfs/task_05_sysfs.c [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+//###########################INIT CALLBACK#####################################
+///////////////////////////////////////////////////////////////////////////////
+
+int __init mod_init(void)
+{
+	int res;
+
+	mod_class = class_create(THIS_MODULE, "task_05");
+	if (IS_ERR(mod_class))
+		pr_info("bad class create\n");
```
Comment (ekovalyov) [inline 05_debug/sysfs/task_05_sysfs.c]: Consider using pr_err and return.

File: 05_debug/sysfs/task_05_sysfs.c [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	res = class_create_file(mod_class, &class_attr_text);
+	res = class_create_file(mod_class, &class_attr_total_calls);
+	res = class_create_file(mod_class, &class_attr_total_characters_processed);
```
Comment (ekovalyov) [inline 05_debug/sysfs/task_05_sysfs.c]: res value is ignored.
Please provide errors handling.

File: 05_debug/sysfs/task_05_sysfs.c [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+ * @param[in]  in         Input string with different cases
+ * @param[in]  shift_dir  The shift dir
+ *
+ * @return     Whole string in the same case.
+ */
+char *change_case(const char *in, enum low_up_t shift_dir)
+{
+	uint16_t L = strlen(in);
+	uint16_t i;
+	char *res = kzalloc(sizeof(*res) * L, GFP_KERNEL);
```
Comment (ekovalyov) [inline 05_debug/sysfs/task_05_sysfs.c]: What is the reason for kzalloc? I suppose kalloc is enough?

Comment (DrCaH4ec) [reply 05_debug/sysfs/task_05_sysfs.c]: I just used to use kzalloc in any case.

Comment (ekovalyov) [reply 05_debug/sysfs/task_05_sysfs.c]: What is the difference between kalloc and kzalloc?

**Review by ekovalyov — APPROVED** [review]

**Pull request merged**